modules/nf-httpfs/src/main/nextflow/file/http/LatchPath.groovy

```diff
@@ -199,9 +199,8 @@ class LatchPath extends XPath {
             try {
                 long bytesWritten = 0
                 byte[] buffer = new byte[downloadChunkSize]
-                int bytesRead
-                while ((bytesRead = inputStream.read(buffer)) != -1) {
-                    ByteBuffer byteBuffer = ByteBuffer.wrap(buffer, 0, bytesRead)
+                while ((buffer = inputStream.readNBytes((int)downloadChunkSize)).length != 0) {
```

Comment:
couple of things:

- `downloadChunkSize` should be declared as an `int`, so that you don't need to cast (is anything depending on it being a `long`?)
- Does `readNBytes` return on `EOF` even if `< downloadChunkSize` bytes are read? want to avoid blocking indefinitely
- Does `readNBytes` allocate a new buffer on every call? 
- This inlining is excessive - would just do `while (true)` and add appropriate `break`s

Reply:
- sg
- It will just return a smaller buffer if < downloadChunkSize. If no bytes left in the input stream, it will return an empty buffer.
  - >This method blocks until the requested number of bytes have been read, end of stream is detected, or an exception is thrown.
- Yea. They have an overloaded version which lets you read into a buffer with an offset.
- sg